provider/openstack/provider.go

```diff
@@ -1039,17 +1039,58 @@ func (e *Environ) StartInstance(args environs.StartInstanceParams) (*environs.St
 		args.InstanceConfig.MachineId,
 	)
 
+	waitForActiveServerDetails := func(
+		client *nova.Client,
+		id string,
+	) (server *nova.ServerDetail, err error) {
+		logger.Debugf("waiting for instance %s to start...", id)
+
+		for {
+			server, err = client.GetServer(id)
+			if server.Status != nova.StatusBuild {
+				break
+			}
+			// We don't want to flood the connection while polling the server waiting
+			// for it to start.
+			logger.Debugf("server has status %s, waiting 10 seconds before polling again...", server.Status)
+			time.Sleep(10 * time.Second)
+		}
+		return server, err
+	}
+
 	tryStartNovaInstance := func(
 		attempts utils.AttemptStrategy,
 		client *nova.Client,
 		instanceOpts nova.RunServerOpts,
 	) (server *nova.Entity, err error) {
+		var message string
 		for a := attempts.Start(); a.Next(); {
 			server, err = client.RunServer(instanceOpts)
-			if err == nil || gooseerrors.IsNotFound(err) == false {
+			if err != nil {
+				break
+			}
+			serverDetail, err := waitForActiveServerDetails(client, server.Id)
+			if err != nil {
+				server = nil
+				break
+			} else if serverDetail.Status == nova.StatusActive {
+				break
+			} else if serverDetail.Status == nova.StatusError {
+				// Perhaps there is an error case where a retry in the same AZ
+				// is a good idea.
+				message = serverDetail.Fault.Message
+				logger.Infof("Instance %q in ERROR state with fault %q", server.Id, message)
+				logger.Infof("Deleting instance %q in ERROR state", server.Id)
+				if err = e.terminateInstances([]instance.Id{instance.Id(server.Id)}); err != nil {
+					logger.Debugf("Failed to delete instance in ERROR state, %q", err)
+				}
+				server = nil
 				break
```

Comment:
why not just do `err = errors.New(serverDetail.Fault.Message)` here, rather than below with the condition?